### PR TITLE
creating sign-up endpoint

### PR DESCRIPTION
**Pull request by f-medina**
## Summary
- User sign-up endpoint

## Known Issues
No testing

## Trello Card
https://trello.com/c/BelrmZRS/35-sign-up

### CODE REVIEW

File: app/controllers/user.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+const { newUser } = require('../services/users');
+
+exports.newUser = (req, res, next) => {
+  const { firstName, lastName, email, password } = req.body;
+  newUser({ firstName, lastName, email, password })
```
Comment (jersonlopez) [inline app/controllers/user.js:5]: Always that you use {} you should return something

File: app/controllers/user.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+const { newUser } = require('../services/users');
+
+exports.newUser = (req, res, next) => {
+  const { firstName, lastName, email, password } = req.body;
```
Comment (jersonlopez) [inline app/controllers/user.js:4]: I think you can restrict the fields by applying destructuring in the service and keep the controller as clear as you can do it, some like this: 
```
exports.myController = (req, res, next) =>  
    myService(req.body)
      .then(...)
      .catch(...)
```

what do you think?

File: app/errors.js [hunk 3]
```diff
@@ -8,3 +8,9 @@ exports.databaseError = message => internalError(message, exports.DATABASE_ERROR
 
 exports.DEFAULT_ERROR = 'default_error';
 exports.defaultError = message => internalError(message, exports.DEFAULT_ERROR);
+
+exports.VALIDATION_ERROR = 'validation_error';
+exports.validationError = message => internalError(message, exports.VALIDATION_ERROR);
+
+exports.MAIL_EXIST_ERROR = 'validation_error';
```
Comment (jersonlopez) [inline app/errors.js:15]: this constant value is equal to `exports.VALIDATION_ERROR`

File: app/errors.js [hunk 4]
```diff
@@ -8,3 +8,9 @@ exports.databaseError = message => internalError(message, exports.DATABASE_ERROR
...
+exports.mailExistError = message => internalError(message, exports.VALIDATION_ERROR);
```
Comment (jersonlopez) [inline app/errors.js:16]: here you are passing a wrong `internalCode` because this error is for existing mails

File: app/models/user.js [hunk 5]
```diff
@@ -1,16 +1,32 @@
...
       password: { type: DataTypes.STRING }
     },
     {
-      timestamps: true
+      timestamps: true,
+      underscored: true
     }
   );
+
+  const encryptPassword = async userData => {
```
Comment (jersonlopez) [inline app/models/user.js:19]: this method could be a helper, what do you think?

File: app/models/user.js [hunk 6]
```diff
@@ -1,16 +1,32 @@
...
+    const salt = await bcrypt.genSalt(15);
```
Comment (jersonlopez) [inline app/models/user.js:20]: You should avoid using the magic numbers, I suggest to you put rounds number in an env var

File: app/models/user.js [hunk 7]
```diff
@@ -1,16 +1,32 @@
...
+    // eslint-disable-next-line require-atomic-updates
+    userData.password = await bcrypt.hash(userData.password, salt);
+    return userData;
+  };
+
+  // eslint-disable-next-line no-unused-vars
+  const matchPassword = password => bcrypt.compare(password, this.password);
```
Comment (jersonlopez) [inline app/models/user.js:27]: same here

File: app/routes.js [hunk 8]
```diff
@@ -1,5 +1,9 @@
 const { healthCheck } = require('./controllers/healthCheck');
+const { newUser } = require('./controllers/user');
+const { validateSchemaAndFail } = require('./middlewares/params_validators');
+const { createUserSchema } = require('./schemas/users');
 
 exports.init = app => {
   app.get('/health', healthCheck);
+  app.post('/new-user', validateSchemaAndFail(createUserSchema), newUser);
```
Comment (jersonlopez) [inline app/routes.js:8]: This endpoint is not using the HTTP rest standard, please read, please search about that and if you have any questions you always can type me by slack

File: app/schemas/users.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  email: {
+    in: ['body'],
+    exists: {
+      errorMessage: 'The email cannot be empty'
+    },
+    isEmail: {
+      errorMessage: 'email invalid'
+    },
+    matches: {
+      options: [/@(\.wolox\.com)|wolox\.(com(\.(ar|co|mx|con|cl)))?/],
```
Comment (jersonlopez) [inline app/schemas/users.js:23]: what happens if in the future wolox opens a new country?

**Review by jersonlopez — CHANGES_REQUESTED** [review]

**Review by jonnyalexbh — APPROVED** [review]

**Pull request merged**